Fix/prevent warning when loading bt

Jad-ELHAJJ opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | N/A |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | N/A |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

This PR resolves the BTCPP warning that is logged when `loadBehaviorTree()` calls `createTreeFromFile()` immediately after `registerTreeFromFile()`

The updated approach ensures that `createTree()` is called directly after `registerTreeFromFile()` when appropriate, thereby eliminating the warning.

To support this, a small interface change was introduced:
- The behavior_tree goal variable can now represent either a BT file path (like before) or a BT ID.
- loadBehaviorTree() will extract a BT ID if the input is an XML file.
- Otherwise, it assumes the input is already a BT ID and directly calls createTree() using that identifier.

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

SteveMacenski commented:
This causes all of the tests related to this test to fail. How does this resolve the warning from https://github.com/ros-navigation/navigation2/pull/5426#issuecomment-3210358335?

Jad-ELHAJJ commented:
@SteveMacenski feel free to review :)

Jad-ELHAJJ commented on nav2_bt_navigator/behavior_trees/follow_point.xml:
```diff
@@ -2,8 +2,8 @@
 -->
 
-<root BTCPP_format="4" main_tree_to_execute="MainTree">
-  <BehaviorTree ID="MainTree">
```
Yes each BT needs to have a unique ID. 
Explanation:
Let's say there is 2 BTs with same ID "MainTree", the function will registerTreeFromFile both BTs but than later, only the last registered MainTree will be createTree since that function relies on the ID instead of the old one that released a warning(createTreeFromFile).

Jad-ELHAJJ replied:
Loading a BT file will, under the hood, extract the BT ID and create the Tree using that ID. This setup requires each BT to have a unique ID. Otherwise, as I mentioned before, if N `MainTree` would be registered at the end the function will create the Tree using the extracted ID; so if same ID `MainTree` is used the last Tree with the `MainTree` will be created. You can verify this with unit test actually, if same ID is used, the `TestAllSuccess` might fail because `navigate_to_pose_w_replanning_and_recovery` has been registered but another Tree with that same ID will be created instead.

SteveMacenski commented:
@Jad-ELHAJJ please undo all of the formatting changes you did to the XML BT files. It looks like there's now 80-100 line diffs on the XML files when all you really need to do is rename the BTs (diff of ~2 lines).

SteveMacenski commented on nav2_behavior_tree/src/behavior_tree_engine.cpp:
```diff
@@ -102,6 +103,56 @@ BehaviorTreeEngine::createTreeFromFile(
+        "Error: Empty filename/BT_ID passed to extractBehaviorTreeID");
+    return "";
+  }
+  if ((!file_or_id.empty() && file_or_id.length() < 4) ||
```
```suggestion
  // This is a behavior tree ID rather than a filepath
  if ((file_or_id.length() < 4) ||
```

That is redundant because if it were empty, it would have triggered the first condition

Jad-ELHAJJ commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp:
```diff
@@ -99,12 +99,18 @@ class BtActionServer
    */
   bool loadBehaviorTree(
-    const std::string & bt_xml_filename = "");
+    const std::string & bt_xml_filename_or_id = "");
```
https://github.com/ros-navigation/docs.nav2.org/pull/772

SteveMacenski reviewed (changes requested):
Small beans, LGTM. The docs PR needs a copy+paste to another file and then we can merge both!

Jad-ELHAJJ commented:
> Small beans, LGTM. The docs PR needs a copy+paste to another file and then we can merge both!

I added test to cover most of the cases and I also fixed the docs PR :)

SteveMacenski commented:
Thanks for the iteration!

mini-1235 commented:
I can actually see the warnings in the latest commit
```
[component_container_isolated-7] [INFO 1759424050.256294767] [bt_navigator]: Activating (on_activate() at /home/mini1235/navigation2/nav2_bt_navigator/src/bt_navigator.cpp:135)
[component_container_isolated-7] WARNING: You executed BehaviorTreeFactory::createTreeFromFile after registerBehaviorTreeFrom[File/Text].
[component_container_isolated-7] This is NOT, probably, what you want to do.
[component_container_isolated-7] You should probably use BehaviorTreeFactory::createTree, instead
[component_container_isolated-7] WARNING: You executed BehaviorTreeFactory::createTreeFromFile after registerBehaviorTreeFrom[File/Text].
[component_container_isolated-7] This is NOT, probably, what you want to do.
[component_container_isolated-7] You should probably use BehaviorTreeFactory::createTree, instead
```

To reproduce, launch
```
ros2 launch nav2_bringup tb4_loopback_simulation_launch.py
```

SteveMacenski commented:
@Jad-ELHAJJ can you verify and let me know if you see this as well still locally too?

Jad-ELHAJJ commented:
> @SteveMacenski I implemented your suggestion. I also added tests that covers most of the cases. So now, the user can use the filepath and get a warning to migrate. The user can also use the BT ID. And in both cases the user will be notified if multiple behavior trees have the same ID.



> @Jad-ELHAJJ can you verify and let me know if you see this as well still locally too?

As mentioned in the [comment above](https://github.com/ros-navigation/navigation2/pull/5494#issuecomment-3285358731), we settled on letting the user use either file or ID. And in the case of file, the user will get a warning.

SteveMacenski commented:
I thought that was only if they had the same ID, not for **all tree files**. Can you clarify?

mini-1235 commented:
I am not sure what those warnings mean, but I think they shouldn't show up when launching nav2's default bringup (tb3,tb4)

SteveMacenski commented:
I agree, the point of this PR was to remove that. If that's still happening for the default Nav2 bringup with the default Nav2 trees which have that ID set as unique for each tree, this feature may need to be reverted.

Jad-ELHAJJ commented:
A fix PR will be open soon to refactor that, making sure that BTCPP warning would never be logged. The final logic will be:

- If a user has conflicting IDs and specifies a file name, only the Behavior Tree (BT) from that file will be loaded. Other trees with the same ID will be ignored. Subtree functionality will fail if subtrees have conflicting IDs.
- If a user has unique IDs (no conflicts) and specifies a file name, all BTs are loaded, and the subtree feature works correctly.
- If a user has unique IDs and specifies a BT ID, all BTs are loaded, and the subtree feature works correctly.
- If a user has conflicting IDs and specifies a BT ID, an error is reported, but the system still attempts to load the BT corresponding to the specified ID.

SteveMacenski commented:
> When using a SubTree, defined in the same file as the main tree, it is now not possible to use a filepath to specify the BT. 

Interesting. This works correctly when they are in different files, correct? I can see that it may be due to extracting the BT IDs from the XMLs in `BtActionServer<ActionT, NodeT>::loadBehaviorTree` for loading. We don't try to register multiple trees from a file when multiple are present. Seems like a fixable issue to register all trees found in the file. 

 @Jad-ELHAJJ thoughts?

> I must also say that the migration guide is really unclear. I thought it is just a new option to load BTs from different directories, but actually one must rename all the BT IDs and add a new param bt_search_directories to use any custom BTs.

I'd appreciate an update to the docs to make it more clear -- I don't have the best sense of what's not clear to you, so you're in a good position to clarify for the next reader :-) 

The renaming of the IDs is a pain, but only if you want to use the IDs. I asked @Jad-ELHAJJ to make sure that you could use it with filenames instead without that and that should work... From the [current version](https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp#L292-L311) of the branch, on first glance it seems like it should work even if you gave the bt_search_directories a bogus path. It'll try to load those but your main XML provided in the action is loaded separately

adivardi commented:
> I'd appreciate an update to the docs to make it more clear -- I don't have the best sense of what's not clear to you, so you're in a good position to clarify for the next reader :-)

Sure, I can do that, though let's solve the loading issue first.
 
> The renaming of the IDs is a pain, but only if you want to use the IDs. I asked @Jad-ELHAJJ to make sure that you could use it with filenames instead without that and that should work...
Oh, is it? I got a lot of warnings about duplicated IDs, so I assume it was a must (I did try with my xml but that has the subTree issue. I didn't try with another xml)

SteveMacenski commented:
I think the duplicate tree IDs should be fine even with warnings as long as you use the file path (not sure though if you include subtrees). Is what you're trying still using subtrees or no?

adivardi commented:
It did have subTrees, but it was running the other file completely.
But the much bigger issue is the subtrees in the same file don't work when specifying a file.
Should I open an issue for this?

mini-1235 commented:
@adivardi can you also provide a reproducible test case? If the author doesn't reply, I can go ahead and fix this in a few days

mini-1235 commented:
> Should I open an issue for this?

@adivardi any updates? 😄
